### PR TITLE
Support parse json expr

### PR DESCRIPTION
**Pull request by xiaocai2333**
/kind improvement

### CODE REVIEW

File: internal/parser/planparserv2/parser_visitor.go [hunk 1]
```diff
@@ -175,7 +178,8 @@ func (v *ParserVisitor) VisitAddSub(ctx *parser.AddSubContext) interface{} {
 		return fmt.Errorf("invalid arithmetic expression, left: %s, op: %s, right: %s", ctx.Expr(0).GetText(), ctx.GetOp(), ctx.Expr(1).GetText())
 	}
 
-	if !typeutil.IsArithmetic(leftExpr.dataType) || !typeutil.IsArithmetic(rightExpr.dataType) {
+	if (!typeutil.IsArithmetic(leftExpr.dataType) && !typeutil.IsJSONType(leftExpr.dataType)) ||
+		(!typeutil.IsArithmetic(rightExpr.dataType) && !typeutil.IsJSONType(rightExpr.dataType)) {
 		return fmt.Errorf("'%s' can only be used between integer or floating expressions", arithNameMap[ctx.GetOp().GetTokenType()])
```
Comment (jiaoew1991) [inline internal/parser/planparserv2/parser_visitor.go:183]: Update the error message together

File: internal/parser/planparserv2/utils.go [hunk 2]
```diff
@@ -124,6 +127,13 @@ func getSameType(a, b schemapb.DataType) (schemapb.DataType, error) {
 	return schemapb.DataType_None, fmt.Errorf("incompatible data type, %s, %s", a.String(), b.String())
 }
 
+func hasJsonField(a, b schemapb.DataType) bool {
```
Comment (jiaoew1991) [inline internal/parser/planparserv2/utils.go:130]: Name it as JSON uniformly, don't use naming such as Json or json.

File: pkg/util/typeutil/schema.go [hunk 3]
```diff
@@ -287,6 +287,15 @@ func IsIntegerType(dataType schemapb.DataType) bool {
 	}
 }
 
+func IsJSONType(dataType schemapb.DataType) bool {
+	switch dataType {
+	case schemapb.DataType_Array, schemapb.DataType_JSON:
```
Comment (jiaoew1991) [inline pkg/util/typeutil/schema.go:292]: Array types are also considered as JSON?

File: tests/integration/json_expr_test.go [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+	"github.com/milvus-io/milvus-proto/go-api/commonpb"
+	"github.com/milvus-io/milvus-proto/go-api/milvuspb"
+	"github.com/milvus-io/milvus-proto/go-api/schemapb"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/funcutil"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/zap"
+)
+
+func TestJsonExpr(t *testing.T) {
```
Comment (jiaoew1991) [inline tests/integration/json_expr_test.go:41]: ditto

File: pkg/util/typeutil/schema.go (hunk 3 above)
Comment (xiaocai2333) [reply pkg/util/typeutil/schema.go:292]: Now json and array should be handled the same.

File: internal/parser/planparserv2/parser_visitor.go (hunk 1 above)
Comment (xiaocai2333) [reply internal/parser/planparserv2/parser_visitor.go]: fixed

File: tests/integration/json_expr_test.go (hunk 4 above)
Comment (xiaocai2333) [reply tests/integration/json_expr_test.go:41]: fixed

**Comment by xiaocai2333** [comment]
/close
see #23804

**Pull request closed without merging**